vhost-device-spi/src/spi.rs

```diff
@@ -0,0 +1,1055 @@
+    /// Holds pointer to userspace buffer for receive data, or null.
+    rx_buf: u64,
+    /// Length of tx and rx buffers, in bytes.
+    len: u32,
+    /// Temporary override of the device's bitrate.
+    speed_hz: u32,
+    /// If nonzero, how long to delay after the last bit transfer
+    /// before optionally deselecting the device before the next transfer.
+    delay_usecs: u16,
+    /// Temporary override of the device's wordsize.
+    bits_per_word: u8,
+    /// True to deselect device before starting the next transfer.
+    cs_change: u8,
+    /// Number of bits used for writing.
+    tx_nbits: u8,
+    /// Number of bits used for reading.
+    rx_nbits: u8,
+    /// If nonzero, how long to wait between words within one
+    /// transfer. This property needs explicit support in the SPI controller,
+    /// otherwise it is silently ignored
+    word_delay_usecs: u8,
+    pad: u8,
+}
+
+/// SPI definitions
+pub(crate) struct SpiTransReq {
+    pub tx_buf: Vec<u8>,
+    pub rx_buf: Vec<u8>,
+    pub trans_len: u32,
+    pub speed_hz: u32,
+    pub mode: u32,
+    pub delay_usecs: u16,
+    pub bits_per_word: u8,
+    pub cs_change: u8,
+    pub tx_nbits: u8,
+    pub rx_nbits: u8,
+    pub word_delay_usecs: u8,
+    pub cs_id: u8,
+}
+
+/// Trait that represents a SPI Device.
+///
+/// This trait is introduced for development purposes only, and should not
+/// be used outside of this crate. The purpose of this trait is to provide a
+/// mock implementation for the SPI driver so that we can test the SPI
+/// functionality without the need of a physical device.
+pub(crate) trait SpiDevice {
+    // Open the device specified by the controller path.
+    fn open(path: &str) -> Result<Self>
+    where
+        Self: Sized;
+
+    // Corresponds to the SPI_IOC_RD_MAX_SPEED_HZ ioctl call.
+    fn max_speed_hz(&self) -> Result<u32>;
+
+    // Corresponds to the SPI_IOC_WR_MAX_SPEED_HZ ioctl call.
+    fn set_max_speed_hz(&self, max_speed_hz: u32) -> Result<()>;
+
+    // Corresponds to the SPI_IOC_RD_BITS_PER_WORD ioctl call.
+    fn bits_per_word(&self) -> Result<u8>;
+
+    // Corresponds to the SPI_IOC_WR_BITS_PER_WORD ioctl call.
+    fn set_bits_per_word(&self, bpw: u8) -> Result<()>;
+
+    // Corresponds to the SPI_IOC_RD_MODE/SPI_IOC_RD_MODE32 ioctl call.
+    fn mode(&self) -> Result<u32>;
+
+    // Corresponds to the SPI_IOC_WR_MODE/SPI_IOC_WR_MODE32 ioctl call.
+    fn set_mode(&self, mode: u32) -> Result<()>;
+
+    // Corresponds to the default ioctl call.
+    fn rdwr(&self, reqs: &mut [SpiIocTransfer]) -> Result<()>;
+
+    // Detect the spi controller supported mode and delay settings
+    fn detect_supported_features(&self) -> Result<VirtioSpiConfig>;
+}
+
+/// A physical SPI device. This structure can only be initialized on hosts
+/// where `/dev/spidevX.Y` is available.
+#[derive(Debug)]
+pub(crate) struct PhysDevice {
+    file: File,
+}
+
+impl SpiDevice for PhysDevice {
+    fn open(path: &str) -> Result<Self> {
```

Comment:
```suggestion
    fn open(path: &Path) -> Result<Self> {
```

Reply:
Now it is:
`fn open(path: &PathBuf) -> Result<Self> {`

Reply:
`PathBuf` implements `AsRef<Path>` so `&PathBuf` means you cannot pass it a `&Path`, but only an owned reference. Whereas passing `&PathBuf` always gets automatically dereferenced by the compiler to `&Path`. Doesn't make a functional difference though in the code.